Fix showing messages from deleted users

simison opened the pull request:
Resolves https://github.com/Trustroots/trustroots/issues/596

Allow reading messages from deleted users by:
- Passing user `_id` from the API
- Opening message thread with that id instead of username
- Loading messages with that _id even when `userFrom` does not resolve to any user _(still broken)_

Todo:
- [x] Fix frontend, might need some refactoring
- [x] Update tests

### Testing

- Create three users and send messages from one to two others
- Delete one of the user's profile
- See if you can still open the message thread
- Ensure for other users, `?userId=` bit doesn't appear in the URLs
- No JS errors in the console when looking at inbox and threads

simison commented on modules/messages/server/controllers/messages.server.controller.js:
```diff
@@ -63,57 +63,88 @@ function sanitizeMessages(messages) {
+ * @param {ObjectId} authenticatedUserId - ID of currently authenticated user
+ * @returns {Array}
+ */
+function sanitizeThreads(threads, authenticatedUserId, callback) {
```
Refactoring `sanitizeThreads()` is prolly better to do outside this PR for clarity.

simison commented on modules/messages/client/controllers/thread.client.controller.js:
```diff
@@ -78,6 +79,12 @@
+        then: function (callback) {
+          callback();
+        }
+      };
```
Just a hack to get things rendering.

simison commented on modules/messages/client/config/messages.client.routes.js:
```diff
@@ -31,8 +31,27 @@
+              deferred.resolve({
+                _id: $stateParams.userId
+              });
+              return deferred.promise;
```
Very much work-in-progress here.

guaka commented:
This works very well on my localhost with my data, tested twice with my own messages, looks good:

<img width="109" alt="image" src="https://user-images.githubusercontent.com/54728/45000722-b54ee780-afc7-11e8-971a-c95712fd9db5.png">

trmendes commented:
I get a console error (a few) saying 
```
Error: message.userFrom is null
messageRead@http://localhost:3000/modules/messages/controllers/thread.client.controller.js?c=58a3ce31283b:339:1
```

guaka commented:
What are the next steps here? I'd like to get rid of that red 2 :)

guaka commented:
Weird stuff when going to another unrelated profile where, empty messages from multiple random other profiles showing up when starting a thread:

<img width="641" alt="image" src="https://user-images.githubusercontent.com/54728/48870317-b16d7d00-ede0-11e8-9f01-b19343055516.png">

(doing this with copy of prod database on local machine)

guaka commented:
same happens with http://localhost:3000/messages/mostusers

also seems to be the consistently the same profiles showing up as senders of the empty messages

and as a bonus it also shows for non-existing usernames (that have never existed)

<img width="829" alt="image" src="https://user-images.githubusercontent.com/54728/48870591-bbdc4680-ede1-11e8-90d7-dc106fc0c07c.png">

hlibbabii commented:
@simison I guess we should reopen this, right? I can also look into it

simison commented:
Sure! The API changes might be salvageable here, but client-side work isn't usable (Angular vs React).

hlibbabii commented:
@simison Could you maybe check out this when you have some time so that I know if I'm moving in the right direction here?

simison commented:
@hlibbabii yap! sorry I missed this — I'll find time to understand the solution asap. Thanks for the nudge and patience!

simison commented:
Works pretty well! :-) 

I can open threads from deleted user 👍  Similarly if they have `shadowban` or `suspended` role 👍 

Some prop warnings in console when user profile is deleted:
<img width="879" alt="Screenshot 2021-02-10 at 13 17 49" src="https://user-images.githubusercontent.com/87168/107503386-95fa1280-6ba2-11eb-94c8-ba3541db5baf.png">

Also note undefined/missing username and link fields:

<img width="663" alt="missing user" src="https://user-images.githubusercontent.com/87168/107503494-b0cc8700-6ba2-11eb-9015-f92bb1c6255e.png">

I think best would be on these cases:
- just show default avatar (done)
- don't link from avatar to profile if there's no user
- instead of nothing, show "Member not available" or something similar in place of name,

I think this would be fine as-is too, so I'm happy to get this merged and do follow-up for those.

hlibbabii commented:
Current ui of the conversation with removed user:

![image](https://user-images.githubusercontent.com/2955794/108192128-2b9e2080-7114-11eb-8a29-0db6e7f7fb6c.png)

simison commented on modules/messages/client/components/ThreadMessage.js:
```diff
@@ -70,30 +70,34 @@ export default function ThreadMessage({ message, user }) {
               {message.userFrom.displayName}
             </a>
+          ) : (
+            <span>Unknown User</span>
```
What do you think about doing "Unknown member" here instead, to be more in line with language across the site.

simison replied:
Oh and needs translation function.

simison replied:
Fixed in https://github.com/Trustroots/trustroots/pull/633/commits/756fd6feb7d636e4546f93bc1d0ea20009f32b35

simison commented on modules/messages/client/components/Thread.component.js:
```diff
@@ -158,24 +159,63 @@ export default function Thread({ user, profileMinimumLength }) {
+      displayName: 'Unknown User',
+      username: null,
+      member: [],
+      languages: [],
```
I wonder if some of this can be avoided by doing checks like `user?.username`, `user?.languages?.length` in the code instead?

simison commented:
Tested deleted users, suspended users — all good, messages get marked as read.

When non-deleted user had written last message in the thread, server-side failed for checking with un-existing `_id`, so I fixed that. Also changed "Unknown User" to slightly friendlier "Unknown member". 

Thanks for helping to get this over finish line @hlibbabii ! 🙌

simison commented:
Except that this doesn't now work for existing users 😁 

![image](https://user-images.githubusercontent.com/87168/109421897-e7e3ca80-79e1-11eb-85dd-bde3b1e8736e.png)

simison commented:
Added small notification to frontend when user is not available:

<img width="586" alt="Screenshot 2021-02-28 at 16 51 45" src="https://user-images.githubusercontent.com/87168/109422718-71e16280-79e5-11eb-956c-64f2f2d420ae.png">
<img width="964" alt="Screenshot 2021-02-28 at 16 51 33" src="https://user-images.githubusercontent.com/87168/109422719-73ab2600-79e5-11eb-9b30-4ce50180b668.png">
<img width="386" alt="Screenshot 2021-02-28 at 16 51 21" src="https://user-images.githubusercontent.com/87168/109422720-74dc5300-79e5-11eb-8acc-b80f50232bc9.png">
